[Wasm-GC] Implement `array.new_canon_fixed`

catamorphism opened the pull request:
#### 3f3a54d2690d395e9171c77d045ef5e183f46438
<pre>
[Wasm-GC] Implement `array.new_canon_fixed`
<a href="https://bugs.webkit.org/show_bug.cgi?id=252350">https://bugs.webkit.org/show_bug.cgi?id=252350</a>

Reviewed by Justin Michaud.

Implement the `array.new_canon_fixed` instruction, which creates a
new array initialized from an arbitrary number of operands (as specified
in the GC proposal MVP).

* JSTests/wasm/gc/array_new_fixed.js: Added.
(module):
(testArrayFuncrefs):
(testArrayExternrefs):
(testArrayRefNull):
(testTypeMismatch):
(testInvalidArrayType):
(testNestedArrays):
(testStructs):
* JSTests/wasm/gc/array_new_fixed_long.js: Added.
(module):
(testLongArray.get let):
(testLongArray):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::getTypeDefinition):
(JSC::Wasm::ExpressionType&gt;::getArrayElementType):
(JSC::Wasm::ExpressionType&gt;::getArrayRefType):
(JSC::Wasm::ExpressionType&gt;::addArrayNew):
(JSC::Wasm::ExpressionType&gt;::addArrayNewDefault):
(JSC::Wasm::ExpressionType&gt;::addArrayNewFixed):
(JSC::Wasm::ExpressionType&gt;::addArrayGet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::getTypeDefinition):
(JSC::Wasm::B3IRGenerator::getArrayTypeDefinition):
(JSC::Wasm::B3IRGenerator::getArrayElementType):
(JSC::Wasm::B3IRGenerator::getArrayRefType):
(JSC::Wasm::B3IRGenerator::addArrayNew):
(JSC::Wasm::B3IRGenerator::addArrayNewDefault):
(JSC::Wasm::B3IRGenerator::addArrayNewFixed):
(JSC::Wasm::B3IRGenerator::addArrayGet):
(JSC::Wasm::B3IRGenerator::emitArrayNullCheck):
(JSC::Wasm::B3IRGenerator::emitArraySetUnchecked):
(JSC::Wasm::B3IRGenerator::addArraySet):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseArrayTypeDefinition):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addArrayNew):
(JSC::Wasm::LLIntGenerator::addArrayNewDefault):
(JSC::Wasm::LLIntGenerator::addArrayNewFixed):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::fillArray):
(JSC::Wasm::arrayNew):
(JSC::Wasm::copyElementsInReverse):
(JSC::Wasm::arrayNewFixed):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/262358@main">https://commits.webkit.org/262358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebe80d1c45d38cb4e6b68abdee2209ee2b86758c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1139 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1170 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1840 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1002 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1250 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1194 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1147 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1053 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1720 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1096 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1051 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1036 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1003 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1040 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2128 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1116 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1087 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1168 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1060 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/237 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/328 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1108 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1171 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/230 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1960,15 +1963,87 @@ auto LLIntGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+    result = push();
+
+    ExpressionType firstValue;
+    shiftOperands(args, firstValue);
```
Hmm, since we end up calling a C thunk in the end anyway, this seems kinda bad. I guess I can't really think of a better way though right now.

Constellation replied:
This looks very dangerous to me. Why not making it LLInt builtin calls?

catamorphism replied:
This follows the structure of `addStructNew()` (or at least it will once https://github.com/WebKit/WebKit/pull/10342 lands), which also calls a C thunk in the end. I was assuming that if there was a better way, it would have been used for struct.new as well (since both take a variable number of arguments), but I'm open to suggestions.

catamorphism replied:
@Constellation I don't see any examples of existing builtins that either take a variable number of arguments or an array of arguments -- how would the arguments be passed if not on the stack?

justinmichaud replied:
I guess the real solution here is that we would normally write this sort of thing directly in the llint, and it would get() them directly. 

As for today, I guess this is fine since it doesn't actually "shift" so much as copy, but it is certainly suboptimal. It shares the same problem as the existing code though, yes.

catamorphism replied:
Waiting for your feedback on #10342 to see whether I should use the same approach here as I just did there.

catamorphism commented on JSTests/wasm/gc/array_new_fixed.js:
```diff
@@ -0,0 +1,599 @@
+testNestedArrays();
+// FIXME: Uncomment when https://bugs.webkit.org/show_bug.cgi?id=252538
+// and https://bugs.webkit.org/show_bug.cgi?id=252719 are fixed
+// testStructs();
```
I wanted to uncomment this line now that these bugs are both fixed, but I found that the test fails if I do; I'm working on it, converted the PR to draft in the meantime.

catamorphism replied:
See https://bugs.webkit.org/show_bug.cgi?id=254226

justinmichaud commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -425,19 +425,28 @@ WASM_SLOW_PATH_DECL(array_new)
+        break;
+    }
+    case Wasm::ArrayGetKind::NewFixed: {
+        ASSERT(!instruction.m_value.isConstant());
```
Why?

catamorphism replied:
Added a comment explaining why.

justinmichaud reviewed (approved):
r=me

catamorphism committed:
[Wasm-GC] Implement `array.new_canon_fixed`
https://bugs.webkit.org/show_bug.cgi?id=252350

Reviewed by Justin Michaud.

Implement the `array.new_canon_fixed` instruction, which creates a
new array initialized from an arbitrary number of operands (as specified
in the GC proposal MVP).

* JSTests/wasm/gc/array_new_fixed.js: Added.
(module):
(testArrayFuncrefs):
(testArrayExternrefs):
(testArrayRefNull):
(testTypeMismatch):
(testInvalidArrayType):
(testNestedArrays):
(testStructs):
* JSTests/wasm/gc/array_new_fixed_long.js: Added.
(module):
(testLongArray.get let):
(testLongArray):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::getTypeDefinition):
(JSC::Wasm::ExpressionType>::getArrayElementType):
(JSC::Wasm::ExpressionType>::getArrayRefType):
(JSC::Wasm::ExpressionType>::addArrayNew):
(JSC::Wasm::ExpressionType>::addArrayNewDefault):
(JSC::Wasm::ExpressionType>::addArrayNewFixed):
(JSC::Wasm::ExpressionType>::addArrayGet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::getTypeDefinition):
(JSC::Wasm::B3IRGenerator::getArrayTypeDefinition):
(JSC::Wasm::B3IRGenerator::getArrayElementType):
(JSC::Wasm::B3IRGenerator::getArrayRefType):
(JSC::Wasm::B3IRGenerator::addArrayNew):
(JSC::Wasm::B3IRGenerator::addArrayNewDefault):
(JSC::Wasm::B3IRGenerator::addArrayNewFixed):
(JSC::Wasm::B3IRGenerator::addArrayGet):
(JSC::Wasm::B3IRGenerator::emitArrayNullCheck):
(JSC::Wasm::B3IRGenerator::emitArraySetUnchecked):
(JSC::Wasm::B3IRGenerator::addArraySet):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseArrayTypeDefinition):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addArrayNew):
(JSC::Wasm::LLIntGenerator::addArrayNewDefault):
(JSC::Wasm::LLIntGenerator::addArrayNewFixed):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::fillArray):
(JSC::Wasm::arrayNew):
(JSC::Wasm::copyElementsInReverse):
(JSC::Wasm::arrayNewFixed):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/262358@main
